hal-x86_64/src/interrupt/mod.rs

```diff
@@ -0,0 +1,419 @@
+        // // (&mut IDT).descriptors[0x20].set_handler(nop as *const ());
+
+        IDT.register_handlers::<TestHandlersImpl>().unwrap();
+        writeln!(&mut writer, "{:#?}", IDT.descriptors[69]).unwrap();
+        IDT.load();
+        IDT.enable();
+    }
+
+    writeln!(&mut writer, "\ttesting interrupts...").unwrap();
+    unsafe {
+        asm!("int $0" :: "i"(69) :: "volatile");
+    }
+    // loop {}
+    writeln!(&mut writer, "\tit worked?").unwrap();
+
+    unsafe { &mut IDT }
+}
+
+impl<'a, T> hal_core::interrupt::Context for Context<'a, T> {
+    type Arch = crate::X64;
+    type Registers = Registers;
+
+    fn registers(&self) -> &Registers {
+        &self.registers
+    }
+
+    /// # Safety
+    ///
+    /// Mutating the value of saved interrupt registers can cause
+    /// undefined behavior.
+    unsafe fn registers_mut(&mut self) -> &mut Registers {
+        &mut self.registers
+    }
+}
+
+impl<'a> hal_core::interrupt::ctx::PageFault for Context<'a, PageFaultCode> {
+    fn fault_vaddr(&self) -> crate::VAddr {
+        unimplemented!("eliza")
+    }
+}
+
+impl<'a> hal_core::interrupt::ctx::CodeFault for Context<'a, ErrorCode> {
+    fn is_user_mode(&self) -> bool {
+        false // TODO(eliza)
+    }
+
+    fn instruction_ptr(&self) -> crate::VAddr {
+        self.registers.instruction_ptr
+    }
+}
+
+impl<'a> Context<'a, ErrorCode> {
+    pub fn error_code(&self) -> ErrorCode {
+        self.code
+    }
+}
+
+impl<'a> Context<'a, PageFaultCode> {
+    pub fn page_fault_code(&self) -> PageFaultCode {
+        self.code
+    }
+}
+
+impl hal_core::interrupt::Control for Idt {
+    type Arch = crate::X64;
+    // type Vector = u8;
+
+    unsafe fn disable(&mut self) {
+        asm!("cli" :::: "volatile");
+    }
+
+    unsafe fn enable(&mut self) {
+        asm!("sti" :::: "volatile");
+    }
+
+    fn is_enabled(&self) -> bool {
+        unimplemented!("eliza do this one!!!")
+    }
+
+    fn register_handlers<H>(&mut self) -> Result<(), hal_core::interrupt::RegistrationError>
+    where
+        H: Handlers<Self::Arch>,
+    {
+        extern "x86-interrupt" fn page_fault_isr<H: Handlers<crate::X64>>(
+            registers: &mut Registers,
+            code: PageFaultCode,
+        ) {
+            H::page_fault(Context { registers, code });
+        }
+
+        extern "x86-interrupt" fn double_fault_isr<H: Handlers<crate::X64>>(
+            registers: &mut Registers,
+            code: u64,
+        ) {
+            H::double_fault(Context { registers, code });
+        }
+
+        extern "x86-interrupt" fn timer_isr<H: Handlers<crate::X64>>(_regs: &mut Registers) {
+            H::timer_tick();
+            unsafe {
+                PIC.end_interrupt(0x20);
+            }
+        }
+
+        extern "x86-interrupt" fn keyboard_isr<H: Handlers<crate::X64>>(_regs: &mut Registers) {
+            H::keyboard_controller();
+            unsafe {
+                PIC.end_interrupt(0x21);
```

Comment:
this is what i meant elsewhere about having interrupt-source-sensitive trampolines. how do you feel about having `Handlers` be hal-crate-specific bundles of traits? ARM's division by zero condition is just a flag bit, for example, where it's an interrupt for x86.

having bundles of interrupt handlers i think means we can pass some composite type through to `CascadedPic::register_handlers` with a `&mut Idt` so it can add entries at the appropriate locations but wait for `Idt` to actually do `lidt`.